tools/paraclu/main.nf

```diff
@@ -11,65 +11,34 @@ process paraclu {
+        overwrite: true,
+        saveAs: { filename ->
+                      if (opts.publish_results == "none") null
+                      else filename }
+
+    container 'quay.io/biocontainers/paraclu:10--he513fc3_0'
 
     input:
+        val opts
         tuple val(sample_id), path(crosslinks)
 
     output:
-        tuple val(sample_id), path("*_peaks.bed.gz"), emit: peaks
+        tuple val(sample_id), path(output_name), emit: peaks
 
     script:
-
-    """
-    #!/usr/bin/env python
-
-    import pandas as pd
-    from subprocess import call
-    import sys
-    import os
-
-    file_in = "${crosslinks}"
-    print(file_in)
-    #file_in_name = file_in.name
-    #print(file_in_name)
-
-    if file_in.endswith('.gz'):
-      file_out = file_in.replace('.bed', '_peaks.bed')
-    else:
-      file_out = file_in.replace('.bed', '_peaks.bed.gz')
-    
-    df_in = pd.read_csv(file_in,
-                        names = ["chrom", "start", "end", "name", "score", "strand"],
-                        header=None, sep='\t')
-
-    df_out = df_in[['chrom', 'strand', 'start', 'score']]
-
-    df_out.sort_values(['chrom', 'strand', 'start'], ascending=[True, True, True], inplace=True)
-
-    paraclu_input = file_in + '.paraclu_input'
-    paraclu_output = file_in + '.paraclu_output'
-
-    df_out.to_csv(paraclu_input, sep='\t', header=None, index=None)
-
-    call(f'paraclu ${params.paraclu_min_value} "{paraclu_input}" | paraclu-cut.sh -l ${params.paraclu_max_cluster_length} -d ${params.paraclu_min_density_increase} > "{paraclu_output}"', shell=True)
-    df_in = pd.read_csv(paraclu_output,
-                        names = ["sequence_name", "strand","start", "end", "number_of_positions",
-                                "sum_of_data_values", "min_density", "max_density"],
-                        header=None, sep='\t')
-    df_in['fourth_column'] = '.'
-    df_out = df_in[['sequence_name', 'start', 'end', 'fourth_column', 'sum_of_data_values', 'strand']]
-    df_out.sort_values(['sequence_name','start', 'end', 'strand'],
-                      ascending=[True, True, True, True], inplace=True)
-    df_out.to_csv(file_out, sep='\t', header=None, index=None)
-    call(f'rm "{paraclu_input}"', shell=True)
-    call(f'rm  "{paraclu_output}"', shell=True)
-
-    """
+        output_name = "${sample_id}_paraclu_peaks_score_${opts.min_cluster_score}.bed.gz"
+
+        // Adding 1 to the start position in the first awk converts to 1-based coordinates
+        // Subtracting 1 from the start position in the second awk converts to 0-based BED format
+
+        """
+        gunzip -c $crosslinks | \
+        awk '{OFS = "\t"}{print \$1, \$6, \$2+1, \$5}' | sort -k1,1 -k2,2 -k3,3n | \
+        paraclu ${opts.min_cluster_score} - | \
+        paraclu-cut -d ${opts.min_density_increase} -l ${opts.max_cluster_length} |
+        awk '{OFS = "\t"}{print \$1, \$3-1, \$4, ".", \$6, \$2}' |
+        sort -k1,1 -k2,2n | gzip > $output_name
+        """
```

Comment:
Perhaps @amchakra or @charlotte-west can check this code, as it's based on their code from the nf-core pipeline. I've accounted for paraclu and BED coordinate conversions, as I think, from reading the manual, that paraclu outputs 1-based coordinates and BED is 0-based.

Reply:
Does this (line 38) need changing? You are correct that paraclu is 1-based and BED 0-based, so to account for this we have used the end position fro the input BED file (i.e. column 3) rather than start position (column 2) which, as the input BED file is always single nucleotide resolution/wide, implicitly does the coordinate conversion.

What do you think @charlotte-west? Might be clearer to be explicit perhaps?

Reply:
Have slept on it, I agree with switch to be more explicit and also because as a module the input BED file may not always be single-nucleotide.